src/main/java/com/minecolonies/coremod/colony/ColonyList.java

```diff
@@ -0,0 +1,166 @@
+package com.minecolonies.coremod.colony;
+
+import com.minecolonies.coremod.util.Log;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+//TODO extend list/collection
+/**
+ * Data structure for storing colonies, optimized for performance.
+ *
+ * @author Colton
+ */
+public final class ColonyList<T extends IColony> implements Iterable<T>
+{
+    private static final int INITIAL_SIZE = 16;
+
+    private Object[] list = new Object[INITIAL_SIZE];
+
+    private List<Integer> nullIndices = new ArrayList<>();
+
+    private int topID = 0;
+
+    private int size = 0;
+
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public T get(int index) {
+        if (index < 1 || index > list.length) {
+            return null;
+        }
+
+        return (T) list[index];
+    }
+
+    @NotNull
+    public Colony create(World world, BlockPos position) {
+        Colony colony = new Colony(getNextColonyID(), world, position);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'colony' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)